File: Python/notes/basics revise/tictactoe.py
import os
import time 
from colorama import Fore, Style
#initializing the game board
game= ['1','2','3','4','5','6','7','8','9']
entries = [Fore.RED+ 'X'+Style.RESET_ALL, Fore.GREEN+'O'+Style.RESET_ALL]
#function to draw board 
def draw_board():
    os.system('clear')
    #if youre on different OS 'clear' might not work idk, but uncomment the line of code below if it doesnt work
    #os.system('cls')
    print("========= Tic-Tac-Toe =========")
    print()
    print(f"           {game[0]} | {game[1]} | {game[2]}   ")
    print("          ---|---|---")
    print(f"           {game[3]} | {game[4]} | {game[5]}   ")
    print("          ---|---|---")
    print(f"           {game[6]} | {game[7]} | {game[8]}   ")
    print()
    print("===============================")
    print()

#function to update board 
def mark_board(choice, mark):
    if game[choice - 1] not in entries:
        if mark == 'X':
            game[choice - 1] = Fore.RED + 'X' + Style.RESET_ALL  # Red color for 'X'
            time.sleep(0.5)
        else:
            game[choice - 1] = Fore.GREEN + 'O' + Style.RESET_ALL  # Green color for 'O'
            time.sleep(0.5)
    else:
        print("\n!! INVALID MOVE !!")
        time.sleep(1.2)

#function that controls game loop 
def check_for_win():
    win_conditions= [
        [0,1,2],
        [3,4,5],
        [6,7,8],
        [0,3,6],
        [1,4,7],
        [2,5,8],
        [0,4,8],
        [6,4,2]
    ]
    
    for condition in win_conditions:
        if (game[condition[0]] == game[condition[1]] == game[condition[2]] ) and ( game[condition[0]] in entries ):
            return 1 #player won
        
    if all(spot in entries for spot in game):
        return  0 #game draw
    
    else:
        return -1 #game continues

# game loop 
player = 1 
game_status = -1 

while game_status == -1:
    
    draw_board()
    player = 1-player #logic to handle player turns 
    
    valid_move= False
    #loop to get valid input
    while not valid_move: 
        try:
            choice = int(input(f"Player {player} ({'X' if player == 0 else 'O'}) make your move (0-9) : "))
            if choice not in range(1,10):
                print("Invalid input , Enter a number between 1-9. ")

            elif game[choice-1] in ['X','O']:
                print("spot already taken , try another move")
            else:
                valid_move= True
                
        except ValueError:
            print("Invalid input , Enter a number between 1-9. ")
    
    mark_board(choice, 'X' if player == 0 else 'O')
    game_status = check_for_win()

draw_board()
if game_status== 1:
    print(f"{'-'*32}")
    print(f"       Player {player + 1} ({'X' if player == 0 else 'O'}) wins!!")
    print(f"{'-'*32}\n")
else:
    print("Game draw!!")
    

File: C/CourseChallenges/tictactoe.c
#include<stdio.h>
#include<stdlib.h>


//drawboard -- get input/check for invalid inputs  -- update board(drawboard) -- check for win 
//if win = false -- turn palyer 2 -- get input -- update board -- check for win.
int choice=0,player=0;

char square[10]={ 'o', '1', '2', '3', '4', '5', '6', '7', '8', '9' }; //1d array which corresponds to each square in 3x3 square
/*
    1 | 2 | 3  
  -------------
    4 | 5 | 6
  -------------
    7 | 8 | 9
*/

int checkForWin();
void displayBoard();
void markBoard(char mark);


int main()
{
    int gameStatus = 0;

    char mark = ' ';

    player = 1;

    do
    {
      displayBoard();

      // change turns
      player = (player % 2) ? 1 : 2;

      // get input
      printf("Player %d, enter a number: ", player);
      scanf("%d", &choice);

      // set the correct character based on player turn
      mark = (player == 1) ? 'X' : 'O';

      // set board based on user choice or invalid choice
      markBoard(mark);

      gameStatus = checkForWin();

      player++;
      system("clear");

    }while (gameStatus == -1);

    if (gameStatus == 1)
        printf("==>\aPlayer %d win ", --player);
    else
        printf("==>\aGame draw");

    return 0;
}
//Check for win:
// 1 -- game is won by either player 
// 0 -- game drawn
// -1 -- game in progress
int checkForWin(){
    int returnvalue=0;
    
    if(square[1]==square[2] && square[2] ==square[3])
        returnvalue =1;
    
    else if (square[4]==square[5] && square[5] ==square[6])
        returnvalue =1;
    
    else if (square[7]==square[8] && square[8] ==square[9])
        returnvalue =1;
    
    else if (square[1]==square[4] && square[4] ==square[7])
        returnvalue =1;

    else if (square[2]==square[5] && square[5] ==square[8])
        returnvalue =1;

    else if (square[3]==square[6] && square[6] ==square[9])
        returnvalue =1;

    else if (square[1]==square[5] && square[5] ==square[9])
        returnvalue =1;

    else if (square[3]==square[5] && square[5] ==square[7])
        returnvalue =1;

    else if (square[1] != '1' && square[2] != '2' && square[3] != '3' &&
        square[4] != '4' && square[5] != '5' && square[6] != '6' && square[7]!= '7' 
        && square[8] != '8' && square[9] != '9')
        returnvalue = 0;

    else
        returnvalue = -1; 
    
    return returnvalue;
    
}

//function that draws board:
void displayBoard(){
    
    printf("\n\nTic Tac Toe\n\n");
    printf("Player 1 (X) -- Player 2 (Y)\n\n");

    printf(" %c | %c | %c \n",square[1],square[2],square[3]);
    printf("-----------\n");
    printf(" %c | %c | %c \n",square[4],square[5],square[6]);
    printf("-----------\n");
    printf(" %c | %c | %c \n",square[7],square[8],square[9]); 
}

//Function that takes input and updates board :
void markBoard(char mark){
    if (choice == 1 && square[1]=='1')
        square[1]= mark;
    
    else if (choice == 2 && square[2]=='2')
        square[2]= mark;

    else if (choice == 3 && square[3]=='3')
        square[3]= mark;

    else if (choice == 4 && square[4]=='4')
        square[4]= mark;

    else if (choice == 5 && square[5]=='5')
        square[5]= mark;

    else if (choice == 6 && square[6]=='6')
        square[6]= mark;
    
    else if (choice == 7 && square[7]=='7')
        square[7]= mark;

    else if (choice == 8 && square[8]=='8')
        square[8]= mark;

    else if (choice == 9 && square[9]=='9')
        square[9]= mark;

    else{
        printf("!!!!INVALID MOVE!!!!");
        player--;
        getchar();

    }
        
}
